File: macos/UniviewCodeBuddyProxy/Sources/Services/UpstreamClient.swift
import Foundation

// ═══════════════════════════════════════════════
// UpstreamClient — 上游 CodeBuddy API HTTP 客户端
// 负责：URLSession 配置、上游头构建、请求执行、
// 空闲超时包装、FNV-128 Machine-Id 生成
// ═══════════════════════════════════════════════

/// 上游请求错误
struct UpstreamError: Error, LocalizedError {
	let statusCode: Int
	let message: String
	var retryAfter: TimeInterval = 0

	var errorDescription: String? {
		return "upstream \(statusCode): \(message)"
	}
}

/// 上游 CodeBuddy API 客户端
final class UpstreamClient {

	// MARK: - 常量

	/// 上游空闲超时（2 分钟）
	private static let idleTimeout: TimeInterval = 120

	/// 最大错误体读取量（1MB）
	private static let maxErrorBodySize = 1 << 20

	/// HTML 标签正则
	private static let htmlTagPattern = try! NSRegularExpression(pattern: "<[^>]*>")

	/// 受保护头（extraHeaders 不可覆盖）
	private static let protectedHeaders: Set<String> = [
		"Authorization", "X-Machine-Id", "X-User-Id",
		"Content-Type", "Host",
	]

	/// 持久化的 Machine-Id（FNV-128a hex）
	let machineId: String

	// MARK: - URLSession

	/// 上游 HTTP 会话（无总超时，30 分钟响应头超时）
	let urlSession: URLSession

	// MARK: - Init

	init(maxConnsPerHost: Int = Defaults.upstreamMaxConnsPerHost) {
		// 生成 Machine-Id
		self.machineId = Self.computeFNV128()

		// 配置 URLSession
		let config = URLSessionConfiguration.ephemeral
		config.timeoutIntervalForRequest = TimeInterval(Defaults.responseHeaderTimeoutSecs) // 30 分钟响应头超时
		config.timeoutIntervalForResource = 0   // 无总超时
		config.httpMaximumConnectionsPerHost = maxConnsPerHost
		config.urlCache = nil
		config.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData

		self.urlSession = URLSession(configuration: config)
	}

	// MARK: - 上游头构建

	/// 构建发送到上游的完整请求头
	func buildUpstreamHeaders(
		model: String,
		intent: UpstreamIntent,
		userId: String,
		machineId: String,
		bearerToken: String
	) -> [String: String] {
		let intentStr = intent.rawValue.isEmpty ? "craft" : intent.rawValue
		let rid = generateRequestID()
		let span = generateSpanID()

		var headers: [String: String] = [
			"Accept":            "text/event-stream",
			"Content-Type":      "application/json",
			"b3":                "\(rid)-\(span)-1-",
			"X-B3-TraceId":      rid,
			"X-B3-ParentSpanId": "",
			"X-B3-SpanId":       span,
			"X-B3-Sampled":      "1",
			"X-Agent-Intent":    intentStr,
			"X-Env-ID":          "production",
			"X-Domain":          Upstream.domain,
			"X-Product":         "SaaS",
			"X-User-Id":         userId,
			"X-Machine-Id":      machineId,
			"X-Request-ID":      rid,
			"X-Conversation-ID": generateRequestID(),
			"X-Session-ID":      generateRequestID(),
			"X-IDE-Type":        "CLI",
			"X-Product-Version": Upstream.productVersion,
			"User-Agent":        Upstream.userAgent,
		]

		if !bearerToken.isEmpty {
			headers["Authorization"] = "Bearer \(bearerToken)"
		}

		return headers
	}

	/// 将客户端提供的 extraHeaders 合并到 base 头中
	func mergeExtraHeaders(base: inout [String: String], extra: [String: String]) {
		for (key, value) in extra {
			if Self.protectedHeaders.contains(key) {
				continue
			}
			if key == "anthropic-beta" {
				if let existing = base[key], !existing.isEmpty {
					base[key] = existing + "," + value
				} else {
					base[key] = value
				}
			} else {
				base[key] = value
			}
		}
	}

	// MARK: - 上游请求

	/// 发送 POST 请求到上游，返回 URLSession 字节流
	func doUpstreamRequest(
		payload: [String: Any],
		headers: [String: String]
	) async throws -> (URLSession.AsyncBytes, HTTPURLResponse) {
		let payloadData = try JSONSerialization.data(withJSONObject: payload)

		guard let url = URL(string: Upstream.baseURL + Upstream.chatURL) else {
			throw UpstreamError(statusCode: 500, message: "invalid upstream URL")
		}

		var request = URLRequest(url: url)
		request.httpMethod = "POST"
		request.httpBody = payloadData

		for (key, value) in headers {
			request.setValue(value, forHTTPHeaderField: key)
		}

		let (bytes, response) = try await urlSession.bytes(for: request)

		guard let httpResponse = response as? HTTPURLResponse else {
			throw UpstreamError(statusCode: 500, message: "invalid response type")
		}

		guard httpResponse.statusCode == 200 else {
			// 读取错误体（最多 1MB）
			var body = Data()
			var count = 0
			for try await byte in bytes {
				if count >= Self.maxErrorBodySize { break }
				body.append(byte)
				count += 1
			}

			let errText = Self.stripHTML(String(data: body, encoding: .utf8) ?? "")

			var retryAfter: TimeInterval = 0
			if httpResponse.statusCode == 429 {
				retryAfter = Self.parseRetryAfterHeader(httpResponse.allHeaderFields)
			}

			throw UpstreamError(
				statusCode: httpResponse.statusCode,
				message: String(errText.prefix(300)),
				retryAfter: retryAfter
			)
		}

		return (bytes, httpResponse)
	}

	/// 执行上游请求并用空闲超时包装响应体
	func doUpstreamRequestWithIdleTimeout(
		payload: [String: Any],
		headers: [String: String]
	) async throws -> (IdleTimeoutLines, HTTPURLResponse) {
		let (bytes, response) = try await doUpstreamRequest(payload: payload, headers: headers)
		let wrapped = IdleTimeoutLines(source: bytes.lines, idleTimeout: Self.idleTimeout)
		return (wrapped, response)
	}

	/// 收集上游所有 chunk（用于非流式请求）
	func collectUpstreamChunks(
		payload: [String: Any],
		bearer: String,
		timeout: TimeInterval
	) -> UpstreamCollectedResult {
		let sem = DispatchSemaphore(value: 0)
		var result = UpstreamCollectedResult()

		Task {
			do {
				let headers = buildUpstreamHeaders(
					model: payload["model"] as? String ?? "",
					intent: .craft,
					userId: "",
					machineId: machineId,
					bearerToken: bearer
				)

				let (bytes, response) = try await doUpstreamRequest(payload: payload, headers: headers)
				result.statusCode = response.statusCode

				guard response.statusCode == 200 else {
					var body = Data()
					var bodyCount = 0
					for try await byte in bytes {
						body.append(byte)
						bodyCount += 1
						if bodyCount >= Self.maxErrorBodySize { break }
					}
					result.errorText = Self.stripHTML(String(data: body, encoding: .utf8) ?? "")
					sem.signal()
					return
				}

				var contentParts: [String] = []
				var promptTokens = 0
				var completionTokens = 0

				for try await line in bytes.lines {
					let trimmed = line.trimmingCharacters(in: .whitespaces)
					guard trimmed.hasPrefix("data: "), trimmed != "data: [DONE]" else { continue }
					let jsonStr = String(trimmed.dropFirst(6))
					guard let data = jsonStr.data(using: .utf8),
						  let chunk = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }

					// 提取 content
					if let choices = chunk["choices"] as? [[String: Any]],
					   let delta = choices.first?["delta"] as? [String: Any],
					   let content = delta["content"] as? String, !content.isEmpty {
						contentParts.append(content)
					}

					// 提取 usage
					if let usage = chunk["usage"] as? [String: Any] {
						promptTokens = usage["prompt_tokens"] as? Int ?? promptTokens
						completionTokens = usage["completion_tokens"] as? Int ?? completionTokens
					}
				}

				result.contentParts = contentParts
				result.promptTokens = promptTokens
				result.completionTokens = completionTokens
			} catch {
				result.statusCode = 500
				result.errorText = error.localizedDescription
			}
			sem.signal()
		}

		let timeoutSecs = Int(timeout)
		_ = sem.wait(timeout: .now() + .seconds(timeoutSecs))

		return result
	}

	/// 获取 Machine-Id
	func getMachineId() -> String {
		return machineId
	}

	// MARK: - FNV-128a Hash

	/// 基于 hostname + homeDir 生成稳定的 FNV-128a 哈希作为 Machine-Id
	private static func computeFNV128() -> String {
		let hostname = Host.current().localizedName ?? "unknown-host"
		let homeDir = FileManager.default.homeDirectoryForCurrentUser.path
		let seed = hostname + "|" + homeDir

		// FNV-128a 参数（与 Go hash/fnv 一致）
		let offsetBasis: [UInt64] = [0x6295c58d, 0x07bb0142, 0x62b82175, 0x6c62272e]
		let prime: [UInt64] = [0x0000013b, 0x00000000, 0x00000000, 0x00000001]

		var hash = offsetBasis

		for byte in seed.utf8 {
			// hash ^= byte (XOR into least significant limb)
			hash[3] ^= UInt64(byte)

			// hash *= prime (128-bit multiplication using 4 x 64-bit limbs, little-endian)
			var carry: UInt64 = 0
			var newHash: [UInt64] = [0, 0, 0, 0]

			for i in 0..<4 {
				var product = hash[i] &* prime[0]
				var lo = product &+ carry
				carry = (product < lo) ? 1 : 0
				carry += (product >> 63) >> 1  // handle overflow
				newHash[i] = lo

				// Cross terms with higher prime limbs
				for j in 1..<4 {
					if i >= j {
						product = hash[i - j] &* prime[j]
						var sum = newHash[i] &+ product
						if sum < newHash[i] { carry &+= 1 }
						newHash[i] = sum
					}
				}
			}

			hash = newHash
		}

		// 转为十六进制字符串（大端序输出）
		let hex = String(format: "%016llx%016llx%016llx%016llx",
						 hash[3], hash[2], hash[1], hash[0])
		return hex
	}

	// MARK: - ID 生成

	/// 生成 16 字节随机请求 ID（32 hex chars）
	private func generateRequestID() -> String {
		var bytes = [UInt8](repeating: 0, count: 16)
		_ = SecRandomCopyBytes(kSecRandomDefault, 16, &bytes)
		return bytes.map { String(format: "%02x", $0) }.joined()
	}

	/// 生成 8 字节随机 Span ID（16 hex chars）
	private func generateSpanID() -> String {
		var bytes = [UInt8](repeating: 0, count: 8)
		_ = SecRandomCopyBytes(kSecRandomDefault, 8, &bytes)
		return bytes.map { String(format: "%02x", $0) }.joined()
	}

	// MARK: - 工具方法

	/// 去除 HTML 标签
	private static func stripHTML(_ s: String) -> String {
		let range = NSRange(s.startIndex..., in: s)
		let result = htmlTagPattern.stringByReplacingMatches(
			in: s,
			options: [],
			range: range,
			withTemplate: ""
		)
		return result.trimmingCharacters(in: .whitespaces)
	}

	/// 解析 Retry-After 头
	private static func parseRetryAfterHeader(_ headers: [AnyHashable: Any]) -> TimeInterval {
		guard let value = headers["Retry-After"] as? String ?? headers["retry-after"] as? String,
			  !value.isEmpty else {
			return 0
		}

		// 尝试解析为秒数
		if let seconds = Double(value), seconds > 0 {
			return min(seconds, 60) // 上限 60s
		}

		// 尝试解析为 HTTP 日期
		let formatter = DateFormatter()
		formatter.dateFormat = "E, dd MMM yyyy HH:mm:ss z"
		if let date = formatter.date(from: value) {
			let interval = date.timeIntervalSinceNow
			return interval > 0 ? min(interval, 60) : 0
		}

		return 0
	}
}

// ═══════════════════════════════════════════════
// IdleTimeoutLines — 2 分钟空闲超时行序列包装器
// ═══════════════════════════════════════════════

/// 带空闲超时的行序列包装
struct IdleTimeoutLines: AsyncSequence {
	typealias Element = String

	let source: AsyncLineSequence<URLSession.AsyncBytes>
	let idleTimeout: TimeInterval

	/// 异步迭代器
	struct Iterator: AsyncIteratorProtocol {
		private let source: AsyncLineSequence<URLSession.AsyncBytes>
		private let idleTimeout: TimeInterval
		private let stream: AsyncStream<String>
		private var streamIterator: AsyncStream<String>.Iterator
		private let backgroundTask: Task<Void, Never>

		init(source: AsyncLineSequence<URLSession.AsyncBytes>, idleTimeout: TimeInterval) {
			self.source = source
			self.idleTimeout = idleTimeout

			let (localStream, continuation) = AsyncStream<String>.makeStream(of: String.self)

			self.stream = localStream
			self.streamIterator = localStream.makeAsyncIterator()

			self.backgroundTask = Task {
				var lastActivity = Date()
				let timeoutTask = Task {
					while !Task.isCancelled {
						try? await Task.sleep(nanoseconds: 1_000_000_000)
						if Date().timeIntervalSince(lastActivity) >= idleTimeout {
							continuation.finish()
							return
						}
					}
				}

				do {
					for try await line in source {
						lastActivity = Date()
						continuation.yield(line)
					}
				} catch {
					// 上游错误，直接结束
				}
				timeoutTask.cancel()
				continuation.finish()
			}
		}

		mutating func next() async -> String? {
			return await streamIterator.next()
		}
	}

	func makeAsyncIterator() -> Iterator {
		Iterator(source: source, idleTimeout: idleTimeout)
	}
}

// ═══════════════════════════════════════════════
// UpstreamCollectedResult — 上游 chunk 收集结果
// ═══════════════════════════════════════════════

struct UpstreamCollectedResult {
	var statusCode: Int = 0
	var errorText: String = ""
	var contentParts: [String] = []
	var promptTokens: Int = 0
	var completionTokens: Int = 0
}

File: windows/UniviewCodeBuddyProxy.Core/Services/UpstreamClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace UniviewCodeBuddyProxy.Services;

/// <summary>Upstream request intent type.</summary>
public enum UpstreamIntent
{
    Craft,
    CodeCompletion,
    Embedding,
}

/// <summary>Upstream request error.</summary>
public sealed class UpstreamError : Exception
{
    public int StatusCode { get; }
    public double RetryAfter { get; }

    public UpstreamError(int statusCode, string message, double retryAfter = 0)
        : base($"upstream {statusCode}: {message}")
    {
        StatusCode = statusCode;
        RetryAfter = retryAfter;
    }
}

/// <summary>Result of collecting upstream chunks for non-streaming requests.</summary>
public sealed class UpstreamCollectResult
{
    public int StatusCode { get; set; }
    public string ErrorText { get; set; } = "";
    public List<string> ContentParts { get; set; } = [];
    public int PromptTokens { get; set; }
    public int CompletionTokens { get; set; }
    public int TotalTokens { get; set; }
    public int CacheReadInputTokens { get; set; }
    public int CacheCreationInputTokens { get; set; }
    public double Credit { get; set; }
}

/// <summary>
/// HttpClient-based upstream request client.
/// No total timeout, 30min response header timeout.
/// </summary>
public sealed class UpstreamClient : IDisposable
{
    // ═══ Constants ═══

    private static readonly TimeSpan IdleTimeout = TimeSpan.FromSeconds(120);
    private const int MaxErrorBodySize = 1 << 20; // 1MB

    private static readonly HashSet<string> ProtectedHeaders = new(StringComparer.OrdinalIgnoreCase)
    {
        "Authorization", "X-Machine-Id", "X-User-Id", "Content-Type", "Host",
    };

    // ═══ State ═══

    private readonly HttpClient _http;
    public string MachineId { get; }

    // ═══ Constructor ═══

    public UpstreamClient(int maxConnsPerHost = Constants.Defaults.UpstreamMaxConnsPerHost)
    {
        MachineId = ComputeFNV128();

        var handler = new SocketsHttpHandler
        {
            MaxConnectionsPerServer = maxConnsPerHost,
            PooledConnectionIdleTimeout = TimeSpan.FromMinutes(5),
            ConnectTimeout = TimeSpan.FromSeconds(30),
        };

        _http = new HttpClient(handler)
        {
            Timeout = Timeout.InfiniteTimeSpan, // No total timeout
        };

        // Set response header timeout via HttpClient.DefaultRequestHeaders is not available,
        // so we use per-request timeout via CancellationTokenSource
    }

    // ═══ Upstream Header Construction ═══

    /// <summary>
    /// Build complete upstream request headers.
    /// </summary>
    public Dictionary<string, string> BuildUpstreamHeaders(
        string model,
        UpstreamIntent intent,
        string userId,
        string machineId,
        string bearerToken)
    {
        var intentStr = intent switch
        {
            UpstreamIntent.Craft => "craft",
            UpstreamIntent.CodeCompletion => "CodeCompletion",
            UpstreamIntent.Embedding => "embedding",
            _ => "craft",
        };
        var rid = GenerateRequestID();
        var span = GenerateSpanID();

        var headers = new Dictionary<string, string>
        {
            ["Accept"] = "text/event-stream",
            ["Content-Type"] = "application/json",
            ["b3"] = $"{rid}-{span}-1-",
            ["X-B3-TraceId"] = rid,
            ["X-B3-ParentSpanId"] = "",
            ["X-B3-SpanId"] = span,
            ["X-B3-Sampled"] = "1",
            ["X-Agent-Intent"] = intentStr,
            ["X-Env-ID"] = "production",
            ["X-Domain"] = Constants.Upstream.Domain,
            ["X-Product"] = "SaaS",
            ["X-User-Id"] = userId,
            ["X-Machine-Id"] = machineId,
            ["X-Request-ID"] = rid,
            ["X-Conversation-ID"] = GenerateRequestID(),
            ["X-Session-ID"] = GenerateRequestID(),
            ["X-IDE-Type"] = "CLI",
            ["X-Product-Version"] = Constants.Upstream.ProductVersion,
            ["User-Agent"] = Constants.Upstream.UserAgent,
        };

        if (!string.IsNullOrEmpty(bearerToken))
            headers["Authorization"] = $"Bearer {bearerToken}";

        return headers;
    }

    /// <summary>
    /// Merge extra headers from client into base headers.
    /// </summary>
    public static void MergeExtraHeaders(Dictionary<string, string> baseHeaders, Dictionary<string, string>? extra)
    {
        if (extra == null) return;

        foreach (var (key, value) in extra)
        {
            if (ProtectedHeaders.Contains(key))
                continue;

            if (key.Equals("anthropic-beta", StringComparison.OrdinalIgnoreCase))
            {
                if (baseHeaders.TryGetValue(key, out var existing) && !string.IsNullOrEmpty(existing))
                    baseHeaders[key] = existing + "," + value;
                else
                    baseHeaders[key] = value;
            }
            else
            {
                baseHeaders[key] = value;
            }
        }
    }

    // ═══ Upstream Request ═══

    /// <summary>
    /// Send a POST request to upstream, returning the response stream.
    /// </summary>
    public async Task<(Stream Stream, HttpResponseMessage Response)> DoUpstreamRequestAsync(
        Dictionary<string, object> payload,
        Dictionary<string, string> headers)
    {
        var url = Constants.Upstream.BaseURL + Constants.Upstream.ChatURL;
        using var request = new HttpRequestMessage(HttpMethod.Post, url);

        var json = JsonSerializer.Serialize(payload);
        request.Content = new StringContent(json, Encoding.UTF8, "application/json");

        foreach (var (key, value) in headers)
            request.Headers.TryAddWithoutValidation(key, value);

        // 30 minute response header timeout
        using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(30));
        var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cts.Token);

        if (response.StatusCode != HttpStatusCode.OK)
        {
            var body = await ReadErrorBodyAsync(response);
            var errText = StripHTML(body);

            double retryAfter = 0;
            if (response.StatusCode == HttpStatusCode.TooManyRequests)
                retryAfter = ParseRetryAfterHeader(response);

            throw new UpstreamError((int)response.StatusCode, errText[..Math.Min(300, errText.Length)], retryAfter);
        }

        var stream = await response.Content.ReadAsStreamAsync();
        return (stream, response);
    }

    /// <summary>
    /// Collect all upstream chunks for non-streaming requests.
    /// </summary>
    public UpstreamCollectResult CollectUpstreamChunks(
        Dictionary<string, object> payload,
        string bearer,
        double timeoutSecs)
    {
        var result = new UpstreamCollectResult();

        try
        {
            var model = payload.TryGetValue("model", out var m) ? m?.ToString() ?? "" : "";
            var headers = BuildUpstreamHeaders(model, UpstreamIntent.Craft, "", MachineId, bearer);

            var task = Task.Run(async () =>
            {
                try
                {
                    var (stream, response) = await DoUpstreamRequestAsync(payload, headers);
                    result.StatusCode = (int)response.StatusCode;

                    if (!response.IsSuccessStatusCode)
                    {
                        using var reader = new StreamReader(stream);
                        var body = await reader.ReadToEndAsync();
                        result.ErrorText = StripHTML(body);
                        return;
                    }

                    using var sr = new StreamReader(stream);
                    string? line;
                    while ((line = await sr.ReadLineAsync()) != null)
                    {
                        var trimmed = line.Trim();
                        if (!trimmed.StartsWith("data: ") || trimmed == "data: [DONE]") continue;

                        var jsonStr = trimmed[6..];
                        try
                        {
                            using var doc = JsonDocument.Parse(jsonStr);
                            var root = doc.RootElement;

                            // Extract content
                            if (root.TryGetProperty("choices", out var choices) && choices.GetArrayLength() > 0)
                            {
                                var firstChoice = choices[0];
                                if (firstChoice.TryGetProperty("delta", out var delta))
                                {
                                    if (delta.TryGetProperty("content", out var contentEl))
                                    {
                                        var content = contentEl.GetString();
                                        if (!string.IsNullOrEmpty(content))
                                            result.ContentParts.Add(content);
                                    }
                                }
                            }

                            // Extract usage
                            if (root.TryGetProperty("usage", out var usage))
                            {
                                if (usage.TryGetProperty("prompt_tokens", out var pt))
                                    result.PromptTokens = pt.GetInt32();
                                if (usage.TryGetProperty("completion_tokens", out var ct))
                                    result.CompletionTokens = ct.GetInt32();
                                if (usage.TryGetProperty("total_tokens", out var tt))
                                    result.TotalTokens = tt.GetInt32();
                                if (usage.TryGetProperty("prompt_tokens_details", out var ptd))
                                {
                                    if (ptd.TryGetProperty("cached_tokens", out var cachedEl))
                                    {
                                        var v = cachedEl.ValueKind == JsonValueKind.Number ? cachedEl.GetInt32() : 0;
                                        if (v > 0) result.CacheReadInputTokens = v;
                                    }
                                    if (ptd.TryGetProperty("cache_creation_tokens", out var cctEl))
                                    {
                                        var v = cctEl.ValueKind == JsonValueKind.Number ? cctEl.GetInt32() : 0;
                                        if (v > 0) result.CacheCreationInputTokens = v;
                                    }
                                }
                                if (usage.TryGetProperty("credit", out var cr) && cr.ValueKind == JsonValueKind.Number)
                                    result.Credit = cr.GetDouble();
                            }
                        }
                        catch
                        {
                            // Skip malformed chunks
                        }
                    }
                }
                catch (Exception ex)
                {
                    result.StatusCode = 500;
                    result.ErrorText = ex.Message;
                }
            });

            task.Wait(TimeSpan.FromSeconds(timeoutSecs));
        }
        catch
        {
            // Timeout
        }

        return result;
    }

    /// <summary>
    /// Fetch models from upstream config endpoint.
    /// </summary>
    public async Task<List<Dictionary<string, string>>> FetchModelsAsync()
    {
        var result = new List<Dictionary<string, string>>();

        // Add hardcoded extra models
        foreach (var (name, ownedBy) in Constants.ExtraModels)
        {
            result.Add(new Dictionary<string, string>
            {
                ["id"] = name,
                ["object"] = "model",
                ["owned_by"] = ownedBy,
            });
        }

        try
        {
            var url = Constants.Upstream.BaseURL + Constants.Upstream.ConfigURL;
            using var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.TryAddWithoutValidation("User-Agent", Constants.Upstream.UserAgent);
            request.Headers.TryAddWithoutValidation("X-Product", "SaaS");
            request.Headers.TryAddWithoutValidation("X-Domain", Constants.Upstream.Domain);

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            using var response = await _http.SendAsync(request, cts.Token);
            var body = await response.Content.ReadAsStringAsync();

            using var doc = JsonDocument.Parse(body);
            var root = doc.RootElement;

            if (root.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
            {
                foreach (var model in data.EnumerateArray())
                {
                    var id = model.TryGetProperty("id", out var idEl) ? idEl.GetString() : null;
                    if (string.IsNullOrEmpty(id)) continue;

                    // Skip if already in extra models
                    if (result.Exists(m => m["id"] == id)) continue;

                    var ownedBy = model.TryGetProperty("owned_by", out var obEl)
                        ? obEl.GetString() ?? Constants.InferOwnedBy(id)
                        : Constants.InferOwnedBy(id!);

                    result.Add(new Dictionary<string, string>
                    {
                        ["id"] = id!,
                        ["object"] = "model",
                        ["owned_by"] = ownedBy,
                    });
                }
            }
        }
        catch
        {
            // Return just extra models on failure
        }

        return result;
    }

    // ═══ FNV-128a Hash ═══

    /// <summary>
    /// Generate a stable FNV-128a hash from hostname + homeDir as Machine-Id.
    /// Matches Go fnv.New128a() behavior.
    /// </summary>
    private static string ComputeFNV128()
    {
        var hostname = Environment.MachineName ?? "unknown-host";
        var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var seed = hostname + "|" + homeDir;

        // FNV-128a parameters
        // offset basis: 0x6c62272e07bb014262b821756295c58d
        // prime: 0x0000000001000000000000000000013b
        var hash = new UInt128(
            upper: 0x6c62272e07bb0142UL | (0x62b821756295c58dUL << 0),
            lower: 0x62b821756295c58dUL
        );

        // Use 4 x UInt64 limbs, little-endian
        var h = new ulong[] { 0x6295c58d, 0x07bb0142, 0x62b82175, 0x6c62272e };
        var prime = new ulong[] { 0x0000013b, 0x00000000, 0x00000000, 0x00000001 };

        foreach (var b in Encoding.UTF8.GetBytes(seed))
        {
            // hash ^= byte
            h[3] ^= b;

            // hash *= prime (128-bit multiplication using 4 x 64-bit limbs)
            var newHash = new ulong[4];
            for (var i = 0; i < 4; i++)
            {
                ulong carry = 0;
                for (var j = 0; j <= i; j++)
                {
                    var product = h[i - j] * prime[j];
                    var lo = product + carry;
                    carry = (product < lo) ? 1uL : 0uL;
                    carry += (product >> 63) >> 1;
                    var sum = newHash[i] + lo;
                    if (sum < newHash[i]) carry++;
                    newHash[i] = sum;
                }
            }

            h = newHash;
        }

        // Convert to hex string (big-endian output)
        return $"{h[3]:x16}{h[2]:x16}{h[1]:x16}{h[0]:x16}";
    }

    // ═══ ID Generation ═══

    internal static string GenerateRequestID()
    {
        var bytes = new byte[16];
        Random.Shared.NextBytes(bytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    internal static string GenerateSpanID()
    {
        var bytes = new byte[8];
        Random.Shared.NextBytes(bytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    // ═══ Utility Methods ═══

    private static async Task<string> ReadErrorBodyAsync(HttpResponseMessage response)
    {
        try
        {
            var body = await response.Content.ReadAsStringAsync();
            return body.Length > MaxErrorBodySize ? body[..MaxErrorBodySize] : body;
        }
        catch
        {
            return "";
        }
    }

    private static string StripHTML(string s)
    {
        if (string.IsNullOrEmpty(s)) return s;
        var result = new StringBuilder();
        var inTag = false;
        foreach (var c in s)
        {
            if (c == '<') inTag = true;
            else if (c == '>') { inTag = false; continue; }
            if (!inTag) result.Append(c);
        }
        return result.ToString().Trim();
    }

    private static double ParseRetryAfterHeader(HttpResponseMessage response)
    {
        var value = response.Headers.RetryAfter?.Delta?.TotalSeconds;
        if (value.HasValue && value.Value > 0)
            return Math.Min(value.Value, 60);

        if (response.Headers.TryGetValues("Retry-After", out var values))
        {
            foreach (var v in values)
            {
                if (double.TryParse(v, out var seconds) && seconds > 0)
                    return Math.Min(seconds, 60);
            }
        }

        return 0;
    }

    public void Dispose()
    {
        _http.Dispose();
    }
}
